# Capstone Notebook 3: Testing Models and evaluation

In this notebook, we will evaluate the various models which we have trained in Notebook 2. 

In [1]:
import numpy as np
import pandas as pd
from PIL import Image
import tensorflow as tf

from keras.models import load_model
from keras.preprocessing import image
from keras.optimizers import Adam

df_test = pd.read_csv('./balanced-one-partition/pneumo_dataset_ITI_rev_clean.tsv' , sep="\t")
y, x, in_channel = 524, 524, 1

Using TensorFlow backend.


In [2]:
gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    try:
        # Restrict TensorFlow to only use the fourth GPU
        tf.config.experimental.set_visible_devices(gpus[0], 'GPU')

        # Currently, memory growth needs to be the same across GPUs
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
    except RuntimeError as e:
        # Memory growth must be set before GPUs have been initialized
        print(e)

1 Physical GPUs, 1 Logical GPUs


In [3]:
from pneumo_data_generator import DataGenerator

data_filter  = df_test['Projection']=='PA'
data_filter &= df_test['Partition']=='te'
df_pneumo_2d_test=df_test[data_filter]
data_generator_test = DataGenerator( df_pneumo_2d_test[data_filter ], y, x, in_channel, batch_size=1,shuffle=False)

C:\Users\Ben\anaconda3\envs\tensorflowgpu\lib\site-packages\ipykernel_launcher.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


In [4]:
df_pneumo_2d_test.shape

(2378, 39)

In [5]:
data=data_generator_test.__getitem__(5)
data2=data_generator_test.__getitem__(6)
img=np.squeeze(data[0], axis=(0,))
img=np.squeeze(img, axis=(2,))*255 #denorm after dropping index arrays
#np.squeeze(data[0], axis=(2,))
img2=np.squeeze(data2[0], axis=(0,))
img2=np.squeeze(img2, axis=(2,))*255

[array([0, 1])]
[array([0, 1])]


In [32]:
ig_treated = Image.fromarray(img)
ig_treated.show()

In [6]:
np.trace(img-img2)

9244.0

## CNN Model Evaluation

In [6]:
# CNN model
CNNmodel= load_model('./models/modelCNN-train.hdf5')

C:\Users\Ben\anaconda3\envs\tensorflowgpu\lib\site-packages\keras\engine\saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
  warnings.warn('No training configuration found in save file: '


In [7]:
predict_CNN = CNNmodel.predict(data_generator_test,steps = len(data_generator_test),verbose=1) #calls getitem
#predict = np.argmax(predict, axis=1)

[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
   1/2378 [..............................] - ETA: 42:35[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
   4/2378 [..............................] - ETA: 11:07[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
   8/2378 [..............................] - ETA: 5:52 [array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
  12/2378 [..............................] - ETA: 4:07[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
  16/2378 [..............................] - ETA: 3:14[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
  20/2378 [..............................] - ETA: 2:42[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
  24/2378 [..............................] - ETA: 2:21[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 

[array([0, 1])]
[array([0, 1])]
 269/2378 [==>...........................] - ETA: 42s[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
 272/2378 [==>...........................] - ETA: 42s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 276/2378 [==>...........................] - ETA: 41s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 280/2378 [==>...........................] - ETA: 41s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 284/2378 [==>...........................] - ETA: 41s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 288/2378 [==>...........................] - ETA: 41s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
 292/2378 [==>...........................] - ETA: 41s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 296/2378 [==>...........................] - ETA: 40s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 300/2378 [==>...........................] - ETA

 545/2378 [=====>........................] - ETA: 32s[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
[array([0, 1])]
 549/2378 [=====>........................] - ETA: 32s[array([1, 0])]
[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
 553/2378 [=====>........................] - ETA: 32s[array([0, 1])]
[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
 557/2378 [======>.......................] - ETA: 32s[array([1, 0])]
[array([0, 1])]
[array([1, 0])]
[array([1, 0])]
 561/2378 [======>.......................] - ETA: 32s[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
[array([0, 1])]
 565/2378 [======>.......................] - ETA: 32s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
 569/2378 [======>.......................] - ETA: 32s[array([0, 1])]
[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
 573/2378 [======>.......................] - ETA: 32s[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
 576/2378 [======>.......................] - ETA: 32s[array([1, 0])]
[array([0, 

 822/2378 [=========>....................] - ETA: 26s[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
 826/2378 [=========>....................] - ETA: 26s[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
 830/2378 [=========>....................] - ETA: 26s[array([1, 0])]
[array([1, 0])]
[array([1, 0])]
 833/2378 [=========>....................] - ETA: 26s[array([1, 0])]
[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
 837/2378 [=========>....................] - ETA: 26s[array([1, 0])]
[array([0, 1])]
[array([1, 0])]
 840/2378 [=========>....................] - ETA: 26s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
 844/2378 [=========>....................] - ETA: 26s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
 848/2378 [=========>....................] - ETA: 26s[array([0, 1])]
[array([1, 0])]
[array([1, 0])]
[array([0, 1])]
 852/2378 [=========>....................] - ETA: 26s[array([0, 1])]
[array([0, 1])]
[array([1, 

1095/2378 [============>.................] - ETA: 21s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1099/2378 [============>.................] - ETA: 21s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1103/2378 [============>.................] - ETA: 21s[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
1107/2378 [============>.................] - ETA: 21s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1110/2378 [=============>................] - ETA: 21s[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
1113/2378 [=============>................] - ETA: 21s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
1117/2378 [=============>................] - ETA: 21s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1121/2378 [=============>................] - ETA: 21s[array([1, 0])]
[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
1125/2378 [=============>................] - ETA: 21s[array([0, 1])]
[array([1, 0])]
[array([1, 

[array([0, 1])]
1372/2378 [================>.............] - ETA: 16s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([1, 0])]
1376/2378 [================>.............] - ETA: 16s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1380/2378 [================>.............] - ETA: 16s[array([0, 1])]
[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
1384/2378 [================>.............] - ETA: 16s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1388/2378 [================>.............] - ETA: 16s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1392/2378 [================>.............] - ETA: 16s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1396/2378 [================>.............] - ETA: 16s[array([1, 0])]
[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1400/2378 [================>.............] - ETA: 16s[array([0, 1])]
[array([0, 1])]
[array([0, 1])]
1403/2378 [================>.............] - ETA: 16s[array([1, 

KeyboardInterrupt: 

In [42]:
predict_CNN

array([[0.21354504, 0.786455  ]], dtype=float32)

In [7]:
df = pd.DataFrame(predict_CNN)

In [11]:
df[0].unique()

array([0.20709012], dtype=float32)

## VGG Model Evaluation

In [8]:
# VGG model
vggmodel= load_model('./models/vggmodel-train.hdf5')

In [9]:
predict_vgg = vggmodel.predict_generator(data_generator_test,steps = len(data_generator_test),verbose=1,
                               workers=1, use_multiprocessing=False)


2378/2378 [==============================] - 68s 28ms/step


In [21]:
predict_vgg

array([[0.20379415, 0.7962058 ],
       [0.20379415, 0.7962058 ],
       [0.20379415, 0.7962058 ],
       ...,
       [0.20379418, 0.7962058 ],
       [0.20379418, 0.7962058 ],
       [0.20379412, 0.7962058 ]], dtype=float32)

## ResNet Model Evaluation

In [9]:
# resnet model
resnetmodel= load_model('./models/resnetmodel-train.hdf5')

In [11]:
predict_resnet = resnetmodel.predict(data_generator_test,steps = len(data_generator_test),verbose=1)


2378/2378 [==============================] - 94s 39ms/step


In [12]:
predict_resnet

array([[0.19495799, 0.805042  ],
       [0.19495799, 0.805042  ],
       [0.19495799, 0.805042  ],
       ...,
       [0.19495799, 0.805042  ],
       [0.19495799, 0.805042  ],
       [0.19495799, 0.805042  ]], dtype=float32)

## Inception Model Evaluation

In [1]:
#inception model
inceptionmodel= load_model('./models/inceptionmodel-train.hdf5')

NameError: name 'load_model' is not defined

In [6]:
predict_inception = inceptionmodel.predict(data_generator_test,steps = len(data_generator_test),verbose=1)


2378/2378 [==============================] - 174s 73ms/step


In [7]:
predict_inception

array([[0.19546415, 0.80453587],
       [0.19546415, 0.80453587],
       [0.19546415, 0.80453587],
       ...,
       [0.19546415, 0.80453587],
       [0.19546415, 0.80453587],
       [0.19546415, 0.80453587]], dtype=float32)

Clearly, all 4 models are pretty close to each other in terms of their accuracy, with little to differentiate between them. This brings us to the next criterion of the models, which is the number of parameters, and the amount taken to train each of the models. They are summarised in the table below:

|Model	|Train Score|Validation Score|Test Score|No. of Params| Training Time|
|---	|---	|---|---|---|---|
|CNN|0.8044|0.8111|0.7930|60,940,898|17min|
|VGG16|0.8044|0.8111|0.7962|15,238,018|32 min|
|ResNet50|0.8041|0.8111|0.8050|25,678,786|47 min|
|InceptionNet|0.8042 |0.8111|0.8045|55,221,090|1h|

While ResNet50 and InceptionNet were marginally better at predicting the test scores, the sheer amount of time they needed to train prohibits them to be used on any but the most powerful computing devices for computation. 

On the other hand, prediction remained relatively computationally inexpensive, with each sample run of 2000+ images in the test set being cleared in less than 3 min. Thus, ResNet50 and InceptionNet may still have some value locally on machines in hospitals, which may make them useful for diagnosis. 

Another factor is the size of the model .hdf5 file as well, which are on the order of megabytes. These may make it difficult for hospitals to retrieve, or store these models locally, especially since there will be models for many other diseases who are using vision machine learning for diagnosis as well.


## Future Work
These models can be further extended to Computed Tomography scans, which are X-ray images done in 3-dimensions. A relatively simple 3layer model was reported to be able to achieve 83% accuracy for a similar pneumonia study:
https://keras.io/examples/vision/3D_image_classification/
